backends/dpdk/dpdkArch.cpp

```diff
@@ -2374,6 +2402,191 @@ const IR::Node* SplitP4TableCommon::postorder(IR::SwitchStatement* statement) {
     return statement;
 }
 
+/* For regular tables, the direct counter array size is same as table size
+ * For learner tables, the direct counter/meter array size is 4 times the table size */
+int CollectDirectCounterMeter::getTableSize(const IR::P4Table * tbl) {
+    int tableSize = dpdk_default_table_size;
+    auto size = tbl->getSizeProperty();
+    if (size)
+        tableSize = size->asUnsigned();
+    return tableSize;
+}
+
+bool CollectDirectCounterMeter::checkMethodCallInAction(
+                const P4::ExternMethod *a, cstring method,
+                cstring &oneInstance, cstring instancename) {
+    bool found = false;
+    cstring externName = a->originalExternType->getName().name;
+    if (externName == "DirectMeter" || externName == "DirectCounter") {
+        auto di = a->object->to<IR::Declaration_Instance>();
+        cstring instanceName = di->name.name;
+        if (a->method->getName().name == method) {
+            if (instancename != "" && instanceName == instancename) {
+                found = true;
+            }
+            if (oneInstance == "")
+            oneInstance = instanceName;
+            // error if more than one count method found with different instance name
+            if (oneInstance != instanceName) {
+                ::error(ErrorType::ERR_UNEXPECTED, "%1% method for different %2% "
+                         "instances (%3% and %4%) called within same action",
+                         method, externName, oneInstance, instanceName);
+                return false;
+            }
+        }
+    }
+    return found;
+}
+
+bool CollectDirectCounterMeter::ifMethodFound(const IR::P4Action *a,
+                        cstring method, cstring instancename) {
+    cstring oneInstance = "";
+    bool found = false;
+    for (auto stmt : a->body->components) {
+        if (auto mcs = stmt->to<IR::MethodCallStatement>()) {
+            auto mi = P4::MethodInstance::resolve(mcs->methodCall, refMap, typeMap);
+            if (auto a = mi->to<P4::ExternMethod>()) {
+                 found = checkMethodCallInAction(a, method, oneInstance, instancename);
+            }
+        }
+        if (auto assn = stmt->to<IR::AssignmentStatement>()) {
+            if (auto m = assn->right->to<IR::MethodCallExpression>()) {
+                auto mi = P4::MethodInstance::resolve(m, refMap, typeMap);
+                if (auto a = mi->to<P4::ExternMethod>()) {
+                    found = checkMethodCallInAction(a, method, oneInstance, instancename);
+                }
+            }
+        }
+    }
+    return found;
+}
+
+void CollectDirectCounterMeter::postorder(const IR::P4Action* a) {
+    ifMethodFound(a, "count");
```

Comment:
you ignore the result, that's strange

Reply:
ifMethodFound function is called from 2 places. 
One for making sure that the default action of owner table of the Direct Meter/Counter extern instance has a execute/count method. In this case, the return value is checked to issue error if specified method call doesn't exist in default action.
Second, for making sure that an action only contains count/execute method calls for only one Direct counter/meter instance. In this case, the error is emitted in the ifMethodFound function itself and hence we are not using the return value here.

Reply:
This would be a good comment